src/serialization.cpp

```diff
@@ -155,10 +168,11 @@ void decompressZlib(std::istream &is, std::ostream &os)
 			zerr(status);
 			throw SerializationError("decompressZlib: inflate failed");
 		}
-		int count = bufsize - z.avail_out;
+		int count = output_size - z.avail_out;
 		//dstream<<"count="<<count<<std::endl;
 		if(count)
 			os.write(output_buffer, count);
+		bytes_written += count;
```

Comment:
couldn't you replace the entire limit impl with a simple
`ìf(limit && (limit -= count) <= 0)`
`    break;`

Reply:
I'd have to unravel the implications of your code, but that's the point – I'd have to unravel it. It modifies a passed parameter and uses `limit == 0` as two different things: that no limit should apply, or (on the right hand side of the operator) that the limit has been reached.

I purposefully left it a bit verbose, since it's a lot clearer this way, in my opinion. The compiler should be able to take care of most inefficiencies.

Reply:
Assignments in if statements are disgusting, especially like that. Shorter code is not usually better code.

Reply:
I'm not saying you should use assignments inside if statements.

What I'm saying is that your implementation could be a lot shorter.
I don't even know why you're modifying `z.avail_out`.

Reply:
Well, the `z.avail_out` is passed to Zlib, which will happily fill all of it (at some CPU cost), and the preexisting code at line 174 pushes everything that comes from zlib onto the output stream. If I keep that, then I would prevent memory DOS attacks, but the function's contract would become "promises not to output _that_ much more than `limit` bytes, say rounded up to the nearest 16K or so, depending on implementation" instead of "promises not to output more than `limit` bytes, no matter what".

And if I only write out the promised part of a full decompressed buffer, people would (rightly) point out that I'm wasting CPU cycles decompressing data that I throw away afterwards.